core/src/main/java/hudson/model/AbstractProject.java

```diff
@@ -2128,7 +2048,9 @@ public void setCustomWorkspace(String customWorkspace) throws IOException {
      * This extension point allows such restrictions.
      *
      * @since 1.540
+     * @deprecated Use {@link LabelExpression.LabelValidator} instead.
      */
+    @Deprecated
```

Comment:
This approach makes it difficult to adopt the new API without bumping the core dependency. E.g. I could not update https://plugins.jenkins.io/label-verifier/ . Maybe it makes sense to introduce a temporary bridge method in this class so that checks could be added in the existing extension point.

Does not block it

Reply:
I mean, the implementation for both validate and auto-complete is standalone.
So it _could_ be created as a separate plugin/library, containing just this functionality (wrapped in some new class like `LabelUtil`, not inside `LabelExpression`). Then new core could use that "plugin", and existing plugins could add a dependency on that rather than a newer Jenkins core.

I'm just not sure a fairly minor feature like this warrants that.

Reply:
It can be done without a plugin and without complex code actually

1. Update LabelValidator

```
public FormValidation checkJob(@NonNull Job<?, ?> job, @NonNull Label label) {
    return FormValidation.ok()
}
```

2. In LabelExpression#validate() add invocation of this method and fail if not OK

3. In a plugin implement a method without adding explicit `@Override` annotation. Thanks to Java, it will still be an override which does not require core bump

Reply:
Ah you meant for the LabelValidator, so for plugins wanting to add validation processing, not plugins wanting to consume the new API.
Sure, I can do that - but I think I'll make the default implementation use `Item`, and call through to the plain `check()` when the passed item is an `AbstractProject`.